src/NuGet.Core/NuGet.Build.Tasks/StaticGraphRestoreTaskBase.cs

```diff
@@ -0,0 +1,306 @@
+        /// </summary>
+        public bool Recursive { get; set; }
+
+        /// <summary>
+        /// Gets or sets the full path to the solution file (if any) that is being built.
+        /// </summary>
+        public string SolutionPath { get; set; }
+
+        protected abstract string DebugEnvironmentVariableName { get; }
+
+        /// <inheritdoc cref="ICancelableTask.Cancel" />
+        public void Cancel() => _cancellationTokenSource.Cancel();
+
+        /// <inheritdoc cref="IDisposable.Dispose" />
+        public void Dispose()
+        {
+            Dispose(disposing: true);
+
+            GC.SuppressFinalize(this);
+        }
+
+        public override bool Execute()
+        {
+            try
+            {
+#if DEBUG
+                var debugRestoreTask = Environment.GetEnvironmentVariable(DebugEnvironmentVariableName);
+                if (!string.IsNullOrEmpty(debugRestoreTask) && debugRestoreTask.Equals(bool.TrueString, StringComparison.OrdinalIgnoreCase))
+                {
+                    Debugger.Launch();
+                }
+#endif
+                DirectoryInfo tempDirectory = Directory.CreateDirectory(NuGetEnvironment.GetFolderPath(NuGetFolderPath.Temp));
+
+                FileInfo responseFile = new FileInfo(Path.Combine(tempDirectory.FullName, Path.GetRandomFileName()));
+
+                try
+                {
+                    WriteResponseFile(responseFile.FullName);
+
+                    MSBuildLogger logger = new MSBuildLogger(Log);
+
+                    using (var semaphore = new SemaphoreSlim(initialCount: 0, maxCount: 1))
+                    using (var loggingQueue = new TaskLoggingQueue(Log))
+                    using (var process = new Process())
+                    {
+                        process.EnableRaisingEvents = true;
+                        process.StartInfo = new ProcessStartInfo
+                        {
+                            Arguments = $"\"{string.Join("\" \"", GetCommandLineArguments(responseFile))}\"",
+                            CreateNoWindow = true,
+                            FileName = GetProcessFileName(ProcessFileName),
+                            RedirectStandardOutput = true,
+                            UseShellExecute = false,
+                            WorkingDirectory = Environment.CurrentDirectory,
+                        };
+
+                        // Place the output in the queue which handles logging messages coming through on StdOut
+                        process.OutputDataReceived += (sender, args) => loggingQueue.Enqueue(args?.Data);
+
+                        process.Exited += (sender, args) => semaphore.Release();
+
+                        try
+                        {
+                            Log.LogMessage(MessageImportance.Low, "Running command: \"{0}\" {1}", process.StartInfo.FileName, process.StartInfo.Arguments);
+
+                            process.Start();
```

Comment:
Unless I am missing something, I don't see a call to `Process.WaitForExit()` method. https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.process.waitforexit?view=net-6.0#system-diagnostics-process-waitforexit(system-int32)

For e.g.  https://github.com/NuGet/NuGet.Client/blob/451713c59f5206f77c97d093dd5191be13f9d0e0/src/NuGet.Core/NuGet.Packaging/NuGetExtractionFileIO.cs#L123

https://github.com/NuGet/NuGet.Client/blob/f24bad0668193ce21a1db8cabd1ce95ba509c7f0/test/NuGet.Core.Tests/NuGet.Common.Test/DirectoryUtilityTests.cs#L83

https://github.com/NuGet/NuGet.Client/blob/451713c59f5206f77c97d093dd5191be13f9d0e0/src/NuGet.Core/NuGet.Packaging/NuGetExtractionFileIO.cs#L123

Reply:
Since another thread is receiving output from the process, WaitForExit() would block the thread.  Instead, I'm using [`EnableRaisingEvents`](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.process.enableraisingevents) and a [`SemaphoreSlim`](https://github.com/NuGet/NuGet.Client/pull/4772/files#diff-bbc3e4c20edf3a20b032ba04bfe636becd616a9455d593f9078e8bc33a33244fR133) which is released when the `Exited` event [fires](https://github.com/NuGet/NuGet.Client/pull/4772/files#diff-bbc3e4c20edf3a20b032ba04bfe636becd616a9455d593f9078e8bc33a33244fR123).